File: src/learning/rule_induction_engine.js
/**
 * Rule Induction Engine
 * Automatically induces new translation rules from post-editing feedback
 */

class RuleInductionEngine {
  constructor() {
    this.learnedRules = []; // Rules learned from post-editing
    this.alignmentModel = null; // Model for aligning source and target words/phrases
    this.patternExtractor = null; // Component to extract linguistic patterns
  }

  /**
   * Induce new rules from post-editing feedback
   * @param {string} originalSource - Original source text
   * @param {string} originalTranslation - Original machine translation
   * @param {string} editedTranslation - Human-edited translation
   */
  induceFromEdit(originalSource, originalTranslation, editedTranslation) {
    console.log('Inducing rules from post-editing feedback...');
    
    // Step 1: Align the original translation with the edited translation
    const alignments = this.alignTranslations(originalTranslation, editedTranslation);
    
    // Step 2: Identify differences between original and edited translations
    const differences = this.identifyDifferences(originalTranslation, editedTranslation, alignments);
    
    // Step 3: Extract patterns from the differences
    const patterns = this.extractPatterns(originalSource, editedTranslation, differences);
    
    // Step 4: Generate new rules based on the patterns
    const newRules = this.generateRules(patterns);
    
    // Step 5: Add the new rules to the learned rules set
    this.learnedRules.push(...newRules);
    
    console.log(`Induced ${newRules.length} new rules from post-editing`);
  }
  
  /**
   * Align original and edited translations
   * @param {string} original - Original translation
   * @param {string} edited - Edited translation
   * @param {Object} alignments - Existing alignments
   * @returns {Array} - Alignment information
   */
  alignTranslations(original, edited, alignments = null) {
    // In a real implementation, this would use sophisticated alignment algorithms
    // like those used in SMT systems (GIZA++, fast_align, etc.)
    
    // For this demo, we'll return a simple placeholder
    return [{
      original: original,
      edited: edited,
      confidence: 0.9
    }];
  }
  
  /**
   * Identify differences between translations
   * @param {string} original - Original translation
   * @param {string} edited - Edited translation
   * @param {Array} alignments - Alignment information
   * @returns {Array} - Differences information
   */
  identifyDifferences(original, edited, alignments) {
    // In a real implementation, this would use diff algorithms
    // and linguistic analysis to identify meaningful differences
    
    // For this demo, we'll return a simple difference
    return [{
      type: 'substitution',
      originalFragment: original,
      editedFragment: edited,
      position: 0
    }];
  }
  
  /**
   * Extract linguistic patterns from differences
   * @param {string} source - Source text
   * @param {string} target - Target translation
   * @param {Array} differences - Differences information
   * @returns {Array} - Extracted patterns
   */
  extractPatterns(source, target, differences) {
    // In a real implementation, this would analyze syntactic and semantic patterns
    // to identify generalizable rules
    
    // For this demo, we'll return a simple pattern
    return [{
      sourcePattern: source,
      targetPattern: target,
      difference: differences[0],
      frequency: 1
    }];
  }
  
  /**
   * Generate new rules from extracted patterns
   * @param {Array} patterns - Extracted patterns
   * @returns {Array} - New rules
   */
  generateRules(patterns) {
    // In a real implementation, this would use rule templating and generalization
    // to create reusable transformation rules
    
    // For this demo, we'll return simple rules based on patterns
    return patterns.map(pattern => ({
      id: `rule_${Date.now()}_${Math.floor(Math.random() * 1000)}`,
      sourcePattern: pattern.sourcePattern,
      targetPattern: pattern.targetPattern,
      condition: 'always',
      confidence: 0.8,
      createdFrom: 'post-editing',
      timestamp: new Date().toISOString()
    }));
  }
  
  /**
   * Get the count of learned rules
   * @returns {number} - Number of learned rules
   */
  getLearnedRuleCount() {
    return this.learnedRules.length;
  }
  
  /**
   * Get all learned rules
   * @returns {Array} - Array of learned rules
   */
  getLearnedRules() {
    return [...this.learnedRules];
  }
  
  /**
   * Clear learned rules
   */
  clearLearnedRules() {
    this.learnedRules = [];
    console.log('Cleared learned rules');
  }
}

module.exports = RuleInductionEngine;

File: src/learning/rule_induction_engine.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Rule induction from post-edit feedback for the Python learning path."""

from __future__ import annotations

from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from difflib import SequenceMatcher
from uuid import uuid4


@dataclass(slots=True)
class InducedRule:
    id: str
    source_pattern: str
    target_pattern: str
    condition: str
    confidence: float
    created_from: str
    timestamp: str
    difference: dict

    def to_dict(self) -> dict:
        return asdict(self)


class RuleInductionEngine:
    """Generate reviewable rule candidates from human edits."""

    def __init__(self):
        self.learned_rules: list[InducedRule] = []

    def induce_from_edit(self, original_source: str, original_translation: str, edited_translation: str) -> list[dict]:
        alignments = self.align_translations(original_translation, edited_translation)
        differences = self.identify_differences(original_translation, edited_translation, alignments)
        patterns = self.extract_patterns(original_source, edited_translation, differences)
        new_rules = self.generate_rules(patterns)
        self.learned_rules.extend(new_rules)
        return [rule.to_dict() for rule in new_rules]

    def align_translations(self, original: str, edited: str, alignments: list[dict] | None = None) -> list[dict]:
        if alignments:
            return alignments
        return [
            {
                "original": original,
                "edited": edited,
                "confidence": round(SequenceMatcher(None, original, edited).ratio(), 3),
            }
        ]

    def identify_differences(self, original: str, edited: str, alignments: list[dict]) -> list[dict]:
        matcher = SequenceMatcher(None, original, edited)
        differences = []
        for tag, i1, i2, j1, j2 in matcher.get_opcodes():
            if tag == "equal":
                continue
            differences.append(
                {
                    "type": tag,
                    "originalFragment": original[i1:i2],
                    "editedFragment": edited[j1:j2],
                    "position": i1,
                }
            )
        return differences or [{"type": "equal", "originalFragment": original, "editedFragment": edited, "position": 0}]

    def extract_patterns(self, source: str, target: str, differences: list[dict]) -> list[dict]:
        return [
            {
                "sourcePattern": source,
                "targetPattern": target,
                "difference": difference,
                "frequency": 1,
            }
            for difference in differences
        ]

    def generate_rules(self, patterns: list[dict]) -> list[InducedRule]:
        timestamp = datetime.now(timezone.utc).isoformat()
        rules = []
        for pattern in patterns:
            rules.append(
                InducedRule(
                    id=f"rule_{uuid4().hex[:12]}",
                    source_pattern=str(pattern.get("sourcePattern") or ""),
                    target_pattern=str(pattern.get("targetPattern") or ""),
                    condition="always",
                    confidence=0.8,
                    created_from="post_editing",
                    timestamp=timestamp,
                    difference=dict(pattern.get("difference") or {}),
                )
            )
        return rules

    def get_learned_rule_count(self) -> int:
        return len(self.learned_rules)

    def get_learned_rules(self) -> list[dict]:
        return [rule.to_dict() for rule in self.learned_rules]

    def clear_learned_rules(self):
        self.learned_rules = []
